### ISSUE TITLE
Handle multiple netlink messages in single response

### ISSUE THREAD

**Issue by gluxon** [jbaublitz/neli#15]
I'm not sure what circumstances this occurs under, but occasionally the kernel can return multiple Netlink messages in the same `libc::recv` call. When this happens, neli interprets the second packet's bytes as an attribute of the first message (which is incorrect).

https://github.com/jbaublitz/neli/blob/1925ba1b3d876f981d0817bbf28ec22b311e6dcf/src/socket.rs#L140-L148

IMHO, `NlSocket::recv` should interpret the first few bytes of the packet, get the length of the next message, and only read that length until the next call to `NlSocket::recv`.

**Comment by jbaublitz** [comment]
Hello! Thanks for opening an issue. I'm just starting back up with this project after not having been able to look at it for a bit.

A few notes:
1. Netlink can indeed have a netlink headers nested within an attribute of another message as is the case with error messages. An error message or an ACK at the protocol level will return a message with a certain `nl_type` indicating whether it is an ACK or error message and will then return the header of the request that generated the error or ACK response. This may appear to be two separate messages but actually is not and the request header is nested under an attribute in the response. Unless you can provide specific scenarios where this is reproduceable, I'm going to assume you're referring to the above scenarios which is defined in the netlink docs.
2. If I am incorrect, the suggestion to only read the first message and then the second is unfortunately not possible with netlink sockets as any data sent in the same "chunk" must be read into a buffer in the same read call or the remaining data that is not read is lost. As I understand it, this is the same as with UDP. As a result, the method of handling netlink messages that I default to is to overestimate the amount of bytes that will come back as the `recv` syscall will handle less bytes being available than the size of the specified buffer but will lose data if underestimated. I've tested this on a number of occasions and this behavior is consistent when reading from a file descriptor into a user space buffer.
3. If point 1 is incorrect, my suggestion for remedying the situation would be, after reading the both messages into a buffer, to simply check if the size of the total parsed netlink message is less than the number of bytes read.

In summary, I do think there would be a parsing error based on the way I handle parsing attributes from the binary representation if this were not indeed one of the situations in point 1 but am open to reproducing and fixing if you can provide me with an example of a scenario where this is not the case. 

Thanks!

**Comment by gluxon** [comment]
> Hello! Thanks for opening an issue. I'm just starting back up with this project after not having been able to look at it for a bit.

Thanks for the response! Didn't expect one so quickly. I appreciate the work you've done on this library and it's definitely been super useful to me.

> 1. Netlink can indeed have a netlink headers nested within an attribute of another message as is the case with error messages. An error message or an ACK at the protocol level will return a message with a certain nl_type indicating whether it is an ACK or error message and will then return the header of the request that generated the error or ACK response. This may appear to be two separate messages but actually is not and the request header is nested under an attribute in the response. Unless you can provide specific scenarios where this is reproduceable, I'm going to assume you're referring to the above scenarios which is defined in the netlink docs.

I wasn't aware of this behavior and you may be right that this is what's happening. Let me try to clarify and see if we're both talking about the same thing though.

I'm trying to receive Netlink multipart messages that have a final `NLMSG_DONE` message. This `NLMSG_DONE` message appears to be consistently included in the same call as the penultimate message. (Irrelevant, but I think this is happening because the kernel module I'm working with has the final done message ready and just ships it in its last call to `send`.) From the [Netlink man page](http://man7.org/linux/man-pages/man7/netlink.7.html), I see that this final message can alternatively be `NLMSG_ERROR`.

Was this the error message you were referring to? I can't find anything on that page that says either of these final message types are included as an attribute rather than its own distinct message. The way I'm seeing the `NLMSG_DONE` formatted through `strace` also makes me think it is its own message. But then again, I'm not very experienced with the Netlink protocol so there may be something else you're referring to that I'm not aware of.

If we're not talking about the same thing, I'll see if I can generate a better example to reproduce.

> 2. If I am incorrect, the suggestion to only read the first message and then the second is unfortunately not possible with netlink sockets as any data sent in the same "chunk" must be read into a buffer in the same read call or the remaining data that is not read is lost. As I understand it, this is the same as with UDP. As a result, the method of handling netlink messages that I default to is to overestimate the amount of bytes that will come back as the recv syscall will handle less bytes being available than the size of the specified buffer but will lose data if underestimated. I've tested this on a number of occasions and this behavior is consistent when reading from a file descriptor into a user space buffer.

In my first comment, I was thinking that you could `MSG_PEEK` a few bytes to get the length and then only read that amount, but I tried implementing this and you were absolutely right! The next message was lost. I wasn't aware that `SOCK_RAW` and `SOCK_DGRAM` behave this way. That definitely makes things more challenging.

> 3. If point 1 is incorrect, my suggestion for remedying the situation would be, after reading the both messages into a buffer, to simply check if the size of the total parsed netlink message is less than the number of bytes read.

Would `NlSocket` have to keep an internal buffer in this case? And read from that on subsequent `recv_nl` calls?

> In summary, I do think there would be a parsing error based on the way I handle parsing attributes from the binary representation if this were not indeed one of the situations in point 1 but am open to reproducing and fixing if you can provide me with an example of a scenario where this is not the case.

I'm not receiving a parsing error, but I am getting the `NLMSG_DONE` payload thrown into an `Unspec` attribute. This happening because the two bytes that correspond to the attribute type are 0's in the same offset of the `NLMSG_DONE` message.

**Comment by jbaublitz** [comment]
Hey @gluxon this follow-up is super helpful so thank you. Let me do a little investigation and refresh my memory on `NLMSG_DONE` and get back to you. This protocol is so large, finicky, and unforgiving it can be really difficult to keep all of the details in my head at the same time sometimes. I've been meaning to write some blog posts on what I'm bumping into for other people in the future who deal with this and so I can look back on it when I inevitably forget.

One thing I can comment on immediately is the frustrating thing I haven't figured out how to address quite yet in my library: the `Unspec` attribute type you mentioned. An unfortunate reality of attribute types in netlink is that sometimes they are used as types and sometimes as an index for a list of nested attributes. I feel this is a really difficult design decision for people writing userspace libraries to work with but it does seem like this is the case based on the results I get and the look I've taken into the kernel code that constructs and sends ACPI events over netlink. The way I try to look at it at a high level is if the kernel is returning a list, it will number the items using the attribute type as an index. If it is returning a map, the attribute type is a key. If you have any suggestions for how to document/approach this from a better angle, I would love to discuss that in a separate issue because I still haven't come up with a good way to make that clear.

**Comment by gluxon** [comment]
> Hey @gluxon this follow-up is super helpful so thank you. Let me do a little investigation and refresh my memory on NLMSG_DONE and get back to you. This protocol is so large, finicky, and unforgiving it can be really difficult to keep all of the details in my head at the same time sometimes. I've been meaning to write some blog posts on what I'm bumping into for other people in the future who deal with this and so I can look back on it when I inevitably forget.

Glad to hear! I do have an example coming, but take your time on anything you need to do. It's definitely been difficult to find documentation on Netlink related things, so happy to hear you might have some blog posts in the future.

> One thing I can comment on immediately is the frustrating thing I haven't figured out how to address quite yet in my library: the Unspec attribute type you mentioned. An unfortunate reality of attribute types in netlink is that sometimes they are used as types and sometimes as an index for a list of nested attributes. I feel this is a really difficult design decision for people writing userspace libraries to work with but it does seem like this is the case based on the results I get and the look I've taken into the kernel code that constructs and sends ACPI events over netlink. The way I try to look at it at a high level is if the kernel is returning a list, it will number the items using the attribute type as an index. If it is returning a map, the attribute type is a key. If you have any suggestions for how to document/approach this from a better angle, I would love to discuss that in a separate issue because I still haven't come up with a good way to make that clear.

Huh. Perhaps I got lucky and the kernel module I'm working with is simple, but I personally worked around this by treating `nla_nested` as an attribute and defining it using `impl_var!`. Then calling `attr.get_attr_handle::<NlaNested>()` to further parse attributes. Not sure if that's representative of what's going on or if it'll work for all generic Netlink families. The perspective I took is that `NlaNested` is a list of attribute types with 1 attribute whose type id is `0`. In this case, I don't see the `0` as `Unspec` but the singleton of `NlaNested`.

**Comment by gluxon** [comment]
Okay, I have a better example. Sorry, I should have tried to provide this from the start so we both knew what was going on better.

## An Example

The following program mimics the `genl ctrl list` command. This prints out all of the generic netlink families available on the host. I used `0.3.1` of this crate, but I'm sure the code on `develop` behaves the same way.

```rust
use neli::err::NlError;
use neli::socket::NlSocket;
use neli::ffi::{CtrlAttr, CtrlCmd, GenlId, NlmF};
use neli::genlhdr::GenlHdr;
use neli::nlhdr::NlHdr;
use neli::nlattr::NlAttrHdr;

fn main() -> Result<(), NlError> {
    let mut socket = NlSocket::<GenlId, GenlHdr<CtrlCmd>>::new_genl()?;

    let attrs: Vec<NlAttrHdr<CtrlAttr>> = vec![];
    let genlhdr = GenlHdr::new(CtrlCmd::Getfamily, 2, attrs)?;
    let nlhdr = NlHdr::new(None, GenlId::Ctrl, vec![NlmF::Request, NlmF::Dump], None, None, genlhdr);
    socket.send_nl(nlhdr)?;

    let response = socket.recv_nl(None)?;
    let mut handle = response.nl_payload.get_attr_handle::<CtrlAttr>();
    handle.parse_nested_attributes()?;

    for attr in handle.iter().unwrap() {
        dbg!(attr);
    }

    Ok(())
}
```

This outputs:

```
[src/main.rs:21] attr = NlAttrHdr {
    nla_len: 11,
    nla_type: FamilyName,
    payload: [
        110,
        108,
        99,
        116,
        114,
        108,
        0,
    ],
}
[src/main.rs:21] attr = NlAttrHdr {
    nla_len: 6,
    nla_type: FamilyId,
    payload: [
        16,
        0,
    ],
}
[src/main.rs:21] attr = NlAttrHdr {
    nla_len: 8,
    nla_type: Version,
    payload: [
        2,
        0,
        0,
        0,
    ],
}
...
[src/main.rs:21] attr = NlAttrHdr {
    nla_len: 156,
    nla_type: Unspec,
    payload: [
        16,
        0,
        2,
        0,
        0,
...
[src/main.rs:21] attr = NlAttrHdr {
    nla_len: 232,
    nla_type: Unspec,
    payload: [
        16,
        0,
        2,
        0,
        0,
        0,
        0,
        0,
        119,
        64,
        0,
        0,
        1,
        2,
        0,
        0,
        15,
...
[src/main.rs:21] attr = NlAttrHdr {
    nla_len: 104,
    nla_type: Unspec,
    payload: [
        16,
        0,
        2,
        0,
        0,
        0,
        0,
        0,
        119,
        64,
        0,
        0,
        1,
        2,
        0,
        0,
        15,
...
```

So the first few attributes detail the `nlctrl` (id `0x10`) family. After the first family is described, all subsequent families appear as an `Unspec` attribute.

I think the fact that each additional family is in its own `Unspec` attribute is by luck. The first `4` bytes of a normal Netlink message is its length. The first `2` bytes of a generic Netlink attribute also describes length. So `2` bytes are consumed by neli to interpret attribute length, and the remaining 2 bytes neli uses to interpret attribute type. However, this is `0x0000` (which is `Unspec`'s id) since most packet sizes are smaller than 2^16. If I wasn't running the example on an x86 little endian processor, I think neli would have had a parsing error upon reading an attribute with length 0.

The start of each `Unspec` payload here is `16`, which corresponds to the Generic Netlink Ctrl family id.

What I would have expected was each call to `nl_recv` return a list response for each family. This isn't the case because the Generic Netlink Controller is returning the entire list of families in a single `recvmsg` call.

Here's an `strace` that shows this:

```
$ strace genl ctrl list 2>&1
...
bind(3, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 0
getsockname(3, {sa_family=AF_NETLINK, nl_pid=17069, nl_groups=00000000}, [12]) = 0
sendto(3, {{len=20, type=0x10 /* NLMSG_??? */, flags=NLM_F_REQUEST|0x300, seq=1555215459, pid=0}, "\x03\x00\x00\x00"}, 20, 0, NULL, 0) = 20
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 1684
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[{{len=116, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0b\x00\x02\x00\x6e\x6c\x63\x74\x72\x6c\x00\x00\x06\x00\x01\x00\x10\x00\x00\x00\x08\x00\x03\x00\x02\x00\x00\x00"...}, {{len=96, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0e\x00\x02\x00\x56\x46\x53\x5f\x44\x51\x55\x4f\x54\x00\x00\x00\x06\x00\x01\x00\x11\x00\x00\x00\x08\x00\x03\x00"...}, {{len=232, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0e\x00\x02\x00\x4e\x4c\x42\x4c\x5f\x4d\x47\x4d\x54\x00\x00\x00\x06\x00\x01\x00\x13\x00\x00\x00\x08\x00\x03\x00"...}, {{len=156, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x11\x00\x02\x00\x4e\x4c\x42\x4c\x5f\x43\x49\x50\x53\x4f\x76\x34\x00\x00\x00\x00\x06\x00\x01\x00\x14\x00\x00\x00"...}, {{len=156, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x11\x00\x02\x00\x4e\x4c\x42\x4c\x5f\x43\x41\x4c\x49\x50\x53\x4f\x00\x00\x00\x00\x06\x00\x01\x00\x15\x00\x00\x00"...}, {{len=232, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0f\x00\x02\x00\x4e\x4c\x42\x4c\x5f\x55\x4e\x4c\x42\x4c\x00\x00\x06\x00\x01\x00\x16\x00\x00\x00\x08\x00\x03\x00"...}, {{len=104, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0f\x00\x02\x00\x61\x63\x70\x69\x5f\x65\x76\x65\x6e\x74\x00\x00\x06\x00\x01\x00\x17\x00\x00\x00\x08\x00\x03\x00"...}, {{len=108, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x12\x00\x02\x00\x74\x68\x65\x72\x6d\x61\x6c\x5f\x65\x76\x65\x6e\x74\x00\x00\x00\x06\x00\x01\x00\x18\x00\x00\x00"...}, {{len=112, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x10\x00\x02\x00\x74\x63\x70\x5f\x6d\x65\x74\x72\x69\x63\x73\x00\x06\x00\x01\x00\x19\x00\x00\x00\x08\x00\x03\x00"...}, {{len=148, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x09\x00\x02\x00\x53\x45\x47\x36\x00\x00\x00\x00\x06\x00\x01\x00\x1a\x00\x00\x00\x08\x00\x03\x00\x01\x00\x00\x00"...}, {{len=112, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0e\x00\x02\x00\x54\x41\x53\x4b\x53\x54\x41\x54\x53\x00\x00\x00\x06\x00\x01\x00\x1b\x00\x00\x00\x08\x00\x03\x00"...}, {{len=112, type=0x10 /* NLMSG_??? */, flags=NLM_F_MULTI, seq=1555215459, pid=17069}, "\x01\x02\x00\x00\x0e\x00\x02\x00\x77\x69\x72\x65\x67\x75\x61\x72\x64\x00\x00\x00\x06\x00\x01\x00\x1c\x00\x00\x00\x08\x00\x03\x00"...}], iov_len=1684}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 1684
...
```

## Other Libraries

I took a look at how `libmnl` solves this problem. It looks like they just return an array of messages.

https://github.com/threatstack/libmnl/blob/006b3bb603ec2ba246c872c8b06ce43d2225dd63/src/nlmsg.c#L166-L185

`libnl` appears to do the same thing: http://www.infradead.org/~tgr/libnl/doc/core.html#_parsing_a_message

## Proposal

Following `libnl` and `libmnl`, I think `NlSocket.recv_nl` and `NlSocket.recv_nl_typed` should also return a vector. Even for single message responses since there doesn't appear to be any guarantees on how many messages are in a response.

The only other alternative I can think of is to buffer additional messages for future `recv_nl` calls, but I personally prefer the above since it's more representative of what's actually happening, gives users more control on how they want to address this, and matches other popular libraries. Just my 2 cents, but I won't have any grievances with how you choose to address this.

I'd be happy to submit a pull request if you'd like the help.

**Comment by jbaublitz** [comment]
Interesting! Thank you for the example. Looking at this, I can confirm this is not the desired behavior and is probably some kind of parsing error. This may indeed be what you originally suggested describing after looking at the debug information and my own code. I can point you to what I found if that's helpful but ultimately this seems to boil down to a combination of factors:
1. The assumption that only one netlink message is available per read call. This assumption does not seem to be correct based on the evidence you've provided and I will confirm once I am able to reproduce. I have done a lot of netlink tracing but the tool I've used does not distinguish at the buffer level, only the message level which led, in part, to this assumption.
2. My parsing strategy for vectors defined in `lib.rs` allows both deserializing all remaining bytes and deserializing a certain number of bytes. Unfortunately, deserializing a genl header will read all remaining bytes in the buffer after the header into an attribute set. This definitely explains how this would happen in your example if there are two messages in the same buffer.

Thank you for putting in the time to track this down and show me the results. I am happy to look at a PR but am also willing to put something together in a PR, have you test it in your use case prior to merging, and then make the fix available. I do think that if this is indeed what is happening, your solution of returning a vector per buffer read from the file descriptor makes the most sense. I ran into a lot of difficulty making netlink strongly typed when originally designing this library and that would be a relatively simple and painless change to fix this while producing the desired results.

**Comment by gluxon** [comment]
> Interesting! Thank you for the example. Looking at this, I can confirm this is not the desired behavior and is probably some kind of parsing error. This may indeed be what you originally suggested describing after looking at the debug information and my own code. I can point you to what I found if that's helpful but ultimately this seems to boil down to a combination of factors:

Thanks. Take your time.

> Thank you for putting in the time to track this down and show me the results. I am happy to look at a PR but am also willing to put something together in a PR, have you test it in your use case prior to merging, and then make the fix available.

I gave an attempt this afternoon and decided to ask you to put together the PR. I'm still a novice at Rust and I think you'd do a much better job.

> I do think that if this is indeed what is happening, your solution of returning a vector per buffer read from the file descriptor makes the most sense.

Glad to hear!

> I ran into a lot of difficulty making netlink strongly typed when originally designing this library and that would be a relatively simple and painless change to fix this while producing the desired results.

You've most certainly done a good job! All of my "why does the API work this way?" questions made sense after understanding Netlink better.

**Comment by jbaublitz** [comment]
Hi @gluxon, trying to play catch up on a few things I let stagnate in my issues before diving into this. I am going to try to get this done in the next week and will report back here on the name of the feature branch once I make it so you can follow along. Thanks for your patience!

**Comment by jbaublitz** [comment]
I'd like to get your input on something. I was thinking that, as opposed to returning a vector for netlink messages, I could provide an iterator-like handling for multiple messages. I bring this up because `Vec`s do heap allocate and while this may not be a problem, there was a previous issue that wanted at least some control over where the memory was allocated. While that was more in the buffer case, I do want to maintain at least some flexibility in terms of whether data is stored on the stack or the heap and if an `Iterator` implementation is provided, collect is always an option. Do you have thoughts on this?

**Comment by gluxon** [comment]
That makes a lot of sense. I don't see the downsides of returning something Iterator-like, and you get the benefits that you mentioned. Thanks!

**Comment by jbaublitz** [comment]
Hi @gluxon, I have pushed most of the work for the change to the `multi-msg-feature` branch. I'm going to add some more testing and will probably have to rework the `tokio` integration which has not been done. However, feel free to take a look and see if it looks reasonable to you.

## Overview
* `recv_nl` has changed to initialize an internal buffer from which messages can be deserialized
* `recv_nl_next`, `recv_nl_next_override` were added to parse messages one at a time corresponding respectively to the old methods `recv_nl`, `recv_nl_typed`
* The `Iterator` trait has been implemented for `NlSocket` but it is important to note that `nl_recv` must be called before using iterator methods or `None` will be returned
  * Some other options for this is to add complexity and provide a separate `.iter()` method that returns a separate `Iterator`-implementing struct or automatically calling `recv_nl` if the internal buffer is not present but I'm not sure I like either of these better 
* `Iterator` methods can only be used when all messages are guaranteed to have the same type as the first deserialization error will cause `None` to be returned and the rest of the buffer to be dropped
  * While not ideal, this is at least provides `.collect()` capabilities in the relevant case (all same-typed messages) and provides a convenient, well known interface for receiving messages in certain scenarios

I will document all of this in the comments before merging but figured I'd give you a chance to try it out and give feedback before I merge.

**Comment by jbaublitz** [comment]
I decided to try the `.iter()` method approach and it's actually a lot cleaner overall so I will probably take that path.

**Comment by gluxon** [comment]
> I decided to try the .iter() method approach and it's actually a lot cleaner overall so I will probably take that path.

I did think it was a bit weird to have `Iterator` defined on `NlSocket` itself. Glad to hear that this approach is cleaner. It does make more sense to me.

Should I wait a bit longer for the `.iter()` approach to be pushed before looking at the `multi-msg-feature` branch?

And thanks for the quick turnaround time on this change!

**Comment by gluxon** [comment]
> Iterator methods can only be used when all messages are guaranteed to have the same type as the first deserialization error will cause None to be returned and the rest of the buffer to be dropped

I'm sure you've thought about this, but what are your thoughts on [`Box`ing each `Nlmsghdr`](https://doc.rust-lang.org/1.30.0/book/second-edition/ch17-02-trait-objects.html#defining-a-trait-for-common-behavior) that's returned? The benefit is that it would allow `.iter()` to return heterogeneous messages, but introduce a runtime cost.

**Comment by jbaublitz** [comment]
Ah yes, I've played around with similar concepts before. The issue here is that you still need to provide a type for the generics. There are very specific rules around this and I've tried everything from trait objects to all kinds of things with generic type constraints. Granted, I've never tried it with this new method and I will do some investigation to see how doable that is. I have tried this other places, however, and that has been a struggle to even get it to compile. I'm updating the branch with fixes for the `stream` feature that pulls in tokio now. I will make a note to investigate the feasibility of adding what you suggested before writing tests and documenting it.

**Comment by jbaublitz** [comment]
Ok, did some investigation. I would have to make the type `Box<Nl>` which is an option but does obfuscate the meaning a bit. Even if I did do it the way you suggest, there would be no way to pass types to each `.next()` call which is what drives the deserialization in the first place when message types differ from each other. I don't think this is doable unfortunately but am continuing on with work on the PR as originally described. I appreciate the idea nonetheless.

**Comment by gluxon** [comment]
> Ok, did some investigation. I would have to make the type Box<Nl> which is an option but does obfuscate the meaning a bit. Even if I did do it the way you suggest, there would be no way to pass types to each .next() call which is what drives the deserialization in the first place when message types differ from each other. I don't think this is doable unfortunately but am continuing on with work on the PR as originally described. I appreciate the idea nonetheless.

Thanks for giving it a shot!

I mentioned it in a comment on the commit, but if I may ask here: I still see Nlmsghdr::deserialize drain the entire buffer. That change hasn't been made yet, right?

**Comment by jbaublitz** [comment]
Not yet, working on it now. It slipped my mind, sorry about that!

**Comment by jbaublitz** [comment]
I've done some thinking about how to address the error idea you mentioned in the case of `.next()` and what I've decided on is this:
* `.next()` will discard the buffer when it reaches an error as will `.nl_next()`
* `.nl_next_override()` will rewind the buffer to the position before the `.read()` call
I think this makes the most sense as `.next()` and `.nl_next()` can never be used to try, fail, and try a different type of deserialization while `.nl_next_override()` could be used to test deserialization and if it fails, try a different type of deserialization.

I'm going to go with this for now and start working on the "reading the entire remaining buffer" problem you mentioned with the above design in place. Still open to feedback until this is merged so feel free to post back here if you have any suggestions.

**Comment by gluxon** [comment]
Oh, I like this.

I don't think it's any better (and I certainly wouldn't change things if you've started implementing the above), but I was thinking of this from the `TryFrom`/`TryInto` angle before our comment. In this scenario, `.next()` and `.nl_next_override()` would just return an `Option`, which could be unwrapped and converted to some `Nl` struct. Just my 2 cents; probably not worth trying since it involves changing some of the type generics in the library and the above works too.

**Comment by jbaublitz** [comment]
Currently blocked by #17

**Comment by jbaublitz** [comment]
Okay, excellent. Pushed the last big change and have done some testing in my project that uses neli, added a program (`acpi_stream`) to neli that debugs ACPI events sent from the kernel, and I'm about to write a test that multiple messages deserialize properly from the same buffer. That'll be done soon so I'm going to have you test it on your end and once you've confirmed it works, I'll close this out and merge the branch to master.

**Comment by gluxon** [comment]
> Okay, excellent. Pushed the last big change and have done some testing in my project that uses neli, added a program (acpi_stream) to neli that debugs ACPI events sent from the kernel, and I'm about to write a test that multiple messages deserialize properly from the same buffer. That'll be done soon so I'm going to have you test it on your end and once you've confirmed it works, I'll close this out and merge the branch to master.

Wonderful. Would you appreciate the example (listing interfaces using genlctrl) listed above incorporated as an example binary? (In addition to the multiple message test you're going to add.)

**Comment by jbaublitz** [comment]
I can definitely add that in!

**Comment by jbaublitz** [comment]
I'm going to merge `multi-msg-feature` and close this. Feel free to comment here if it's not resolved and I'll reopen. Thanks!

**Comment by gluxon** [comment]
@jbaublitz I think this broke again after commits a614a08 to c58fac2.

In most recent commit at top order:
- c58fac2f -- running, but incorrect response
- 7657225 -- crash
- a614a08 -- crash
- cbb0b41 -- working

When I run `cargo run --example ctrl-list` on `master`, I get:

```
nlctrl
	ID: 0x10
```

The expected output is:

```
nlctrl
	ID: 0x10
VFS_DQUOT
	ID: 0x11
NLBL_MGMT
	ID: 0x13
NLBL_CIPSOv4
	ID: 0x14
NLBL_CALIPSO
	ID: 0x15
NLBL_UNLBL
	ID: 0x16
acpi_event
	ID: 0x17
thermal_event
	ID: 0x18
tcp_metrics
	ID: 0x19
SEG6
	ID: 0x1a
TASKSTATS
	ID: 0x1b
```

I'm seeing that response messages after the first one are being parsed as unrecognized variant attributes of the first again. See: https://github.com/jbaublitz/neli/issues/15#issuecomment-482920458

**Comment by jbaublitz** [comment]
I added a test to prevent this in the future and manual testing of `ctrl-list` seems to work again. I'm going to close this out but master is good to go. Thanks for the report!
